modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeCrystalPlasticityThermalEigenstrain.C

```diff
@@ -0,0 +1,55 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ComputeCrystalPlasticityThermalEigenstrain.h"
+
+registerMooseObject("TensorMechanicsApp", ComputeCrystalPlasticityThermalEigenstrain);
+
+InputParameters
+ComputeCrystalPlasticityThermalEigenstrain::validParams()
+{
+  InputParameters params = ComputeCrystalPlasticityEigenstrainBase::validParams();
+
+  params.addCoupledVar("temperature", "Coupled temperature variable");
+  params.addParam<std::vector<Real>>(
+      "thermal_expansion_coefficients",
+      "Vector of values defining the constant second order thermal expansion coefficients");
+
+  return params;
+}
+
+ComputeCrystalPlasticityThermalEigenstrain::ComputeCrystalPlasticityThermalEigenstrain(
+    const InputParameters & parameters)
+  : DerivativeMaterialInterface<ComputeCrystalPlasticityEigenstrainBase>(parameters),
+    _temperature(coupledValue("temperature")),
+    _temperature_old(coupledValueOld("temperature")),
+    _ddeformation_gradient_dT(declarePropertyDerivative<RankTwoTensor>(
+        _deformation_gradient_name, getVar("temperature", 0)->name()))
+{
+  std::vector<Real> therm_exp_coeff = getParam<std::vector<Real>>("thermal_expansion_coefficients");
+  if (therm_exp_coeff.size() != _mesh.dimension())
+    paramError("thermal_expansion_coefficients",
+               "thermal expansion coefficient size must match the mesh dimension");
+  _thermal_expansion_coefficients.fillFromInputVector(therm_exp_coeff);
```

Comment:
is `_thermal_expansion_coefficients` always a diagonal matrix and filled by three entries?

Reply:
Yeah, this is what I've seen so far from the paper.

Reply:
I just took a look at the `fillFromInputVector` method. It can take the vector size of 1, 3, 6 and 9. If your mesh is 2, you probably still need to provide 3 entries, although I am not exactly sure what the third entry will be.

Reply:
I don't think it is always diagonal. People usually measure the expansion coefficients `a1`, `a2`, `a3` along three principal axes. The only case this would be diagonal is when the principal axes coincide with the cartesian coords.

Why can't we just get a `RankTwoTensor` material property as the `_thermal_expansion_coefficients`? That way the user has freedom to specify whatever anisotropy in their expansion coefficients.

Reply:
@hugary1995 I think what @dewenyushu did here is fine. She basically fill in a tensor with a vector. If we do not limit the size of the vector, then users have freedom to specify  whatever tensor they  want.

Reply:
@jiangwen84 Yeah, it totally makes sense now to remove the limit of the size: 1) it does not make sense in 2D, and 2) as @hugary1995 pointed out there is potential need for off diagonal entries in the thermal expansion coefficient